src/main/java/net/fabricmc/mappingio/Main.java

```diff
@@ -14,118 +14,135 @@
-			mappingTree.accept(writer, VisitOrder.createByName());
-			writer.close();
-		} catch (IOException e) {
-			e.printStackTrace();
-			return 1;
+	public static boolean deleteOutputDirectory(Path directory) {
+		if (!Files.exists(directory)) return true;
+		try {
+			Files.walkFileTree(directory, new SimpleFileVisitor<Path>() {
+				@Override
+				public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+					Files.delete(file);
+					return FileVisitResult.CONTINUE;
+				}
+
+				@Override
+				public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+					Files.delete(dir);
+					return FileVisitResult.CONTINUE;
+				}
+			});
+			return true;
+		} catch (IOException | SecurityException e) {
+			System.out.println("Error when trying to delete " + directory + ": " + e.getMessage());
+			return false;
 		}
-
-		return 0;
 	}
 
-	public static int yarnfulldescs(Path jarInput, Path yarnInput, Path output, MappingFormat outputFormat) {
+	public static int merge(Path primary, Path secondary, Path output, MappingFormat format) {
 		try {
-			MemoryMappingTree mappingTree = new MemoryMappingTree();
-			MemoryMappingTree mappingTreeForward;
-
-			// load jar
-			System.out.println(String.format("reading %s", jarInput));
-
-			{
-				MappingReader.read(jarInput, mappingTree);
-
-				// rename source namespace to official
-				mappingTree.setSrcNamespace("official");
+			System.out.println("Deleting " + output);
+			if (!deleteOutputDirectory(output)) {
+				return 1;
 			}
 
-			// load intermediary and yarn
-			System.out.println(String.format("reading %s", yarnInput));
+			System.out.printf("Reading %s%n", primary);
+			MemoryMappingTree primaryMappingTree = new MemoryMappingTree();
+			MappingReader.read(primary, primaryMappingTree);
+
+			System.out.printf("Reading %s%n", secondary);
+			MemoryMappingTree secondaryMappingTree = new MemoryMappingTree();
+			MappingReader.read(secondary, secondaryMappingTree);
+
+			for (MappingTree.ClassMapping secondaryClass : secondaryMappingTree.getClasses()) {
+				MappingTree.ClassMapping correspondingPrimaryClass = primaryMappingTree.getClass(secondaryClass.getSrcName());
+				if (correspondingPrimaryClass == null) {
+					primaryMappingTree.addClass(secondaryClass);
+				} else {
+					MappingMerger.mergeClassMappingInplace(correspondingPrimaryClass, secondaryClass);
+				}
+			}
 
-			{
-				// filter out anything not in jar
-				MappingReader.read(yarnInput, new MissingSrcFilter(mappingTree));
+			Map<String, List<String>> duplicateMappings = MappingMerger.getDuplicateClassMappings(primaryMappingTree);
+			for (Map.Entry<String, List<String>> entry : duplicateMappings.entrySet()) {
+				System.out.println(entry.getValue().size() + " classes are mapped to '" + entry.getKey() + "': " + String.join(", ", entry.getValue()));
+				for (String srcName : entry.getValue()) {
+					primaryMappingTree.removeClass(srcName);
+				}
```

Comment:
il faudrait une méthode removeAll(Collection) dans MemoryMappingTree pour éviter la complexité en O(n²)

Reply:
Je vois pas le O(n²), pour moi c'est du O(n) avec `n` le nombre de classes dans le `.jar` obfusqué. Et le `removeAll` finirai par faire une boucle `for` aussi non?